extension-manifest-v3/src/background/reporting.js

```diff
@@ -216,23 +216,41 @@ const reporting = new Reporting({
   communication,
 });
 
-observe('terms', (terms) => {
+function delay(timeInMs) {
+  return new Promise((resolve) => setTimeout(resolve, timeInMs));
+}
+
+function defer() {
+  const deferred = {};
+  deferred.promise = new Promise((resolver) => {
+    deferred.resolve = resolver;
+  });
+  deferred.waitUntil = async ({ timeout, errorMessage = '' } = {}) =>
+    Promise.race([
+      deferred.promise,
+      delay(timeout).then(() => Promise.reject(errorMessage)),
+    ]);
+  return deferred;
+}
+
+const deferredStartup = defer();
+
+observe('terms', async (terms) => {
   if (terms) {
-    reporting.init().catch((e) => {
+    try {
+      await reporting.init();
+    } catch (e) {
       console.warn(
         'Failed to initialize reporting. Leaving the module disabled and continue.',
         e,
       );
-    });
+    }
   } else {
     reporting.unload();
   }
+  deferredStartup.resolve(reporting.isActive);
 });
 
-function delay(timeInMs) {
-  return new Promise((resolve) => setTimeout(resolve, timeInMs));
-}
-
 function onLocationChange(details) {
   if (!reporting.isActive) return;
```

Comment:
This checked needs to be removed to resolve the race. Or at least delayed before the call to reporting.analyzeUrl.

Reply:
I will remove it from here as this way all events prior to initialisation of reporting will be dropped.